p2p/src/sync/peer.rs

```diff
@@ -390,29 +390,44 @@ where
             )));
         }
 
-        if let Some(tx) = self
+        let res = if let Some(tx) = self
             .mempool_handle
             .call_async(move |m| Box::pin(async move { m.transaction(&id).await }))
             .await??
         {
-            self.messaging_handle
-                .send_message(self.id(), SyncMessage::TransactionResponse(tx))?;
-        }
+            TransactionResponse::Found(tx)
+        } else {
+            TransactionResponse::NotFound(id)
+        };
+        self.messaging_handle
+            .send_message(self.id(), SyncMessage::TransactionResponse(res))?;
 
         Ok(())
     }
 
-    async fn handle_transaction_response(&mut self, tx: SignedTransaction) -> Result<()> {
-        let id = tx.serialized_hash();
+    async fn handle_transaction_response(&mut self, resp: TransactionResponse) -> Result<()> {
+        match resp {
```

Comment:
The repetition inside match bothers me. 🙃 Perhaps the following would be better?..
```rust
let (id, tx) = match resp {
    TransactionResponse::NotFound(id) => (id, None),
    TransactionResponse::Found(tx) => (tx.serialized_hash(), tx),
};

if self.announced_transactions.take(&id).is_none() {
    return Err(P2pError::ProtocolError(ProtocolError::UnexpectedMessage(
        "Unexpected transaction response",
    )));
}

if let Some(tx) = tx {
    self.mempool_handle.call_async_mut(|m| m.add_transaction(tx)).await??;
    self.messaging_handle.make_announcement(Announcement::Transaction(id))
}
```

Reply:
Btw, we should never use serialized_hash. Always only use get_id.

Reply:
Btw, we should never use serialized_hash. Always only use get_id.a

Reply:
@TheQuantumPhysicist Perhaps I'm missing something, but as far as I can see, `SignedTransaction` doesn't implement the `Idable` trait.

Reply:
Take it from `Transaction` within it. This glosses on the issue of transaction malleability.

Reply:
I updated all places in p2p where `serialized_hash` was used. And I changed the type of `SignedTransaction::serialized_hash` to catch such errors:
```
-    pub fn serialized_hash(&self) -> Id<Transaction> {
+    pub fn serialized_hash(&self) -> Id<SignedTransaction> {
```
Is it OK now?

Reply:
No. Please don't change SignedTransaction

So something like tx.transaction().get_id()

Reply:
OK, I reverted changing `SignedTransaction::serialized_hash` output type.